File: js/ai/tests/formats/json_test.ts
/**
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { z } from '@genkit-ai/core';
import { Registry } from '@genkit-ai/core/registry';
import assert from 'node:assert';
import { beforeEach, describe, it } from 'node:test';
import { configureFormats } from '../../src/formats/index.js';
import { jsonFormatter } from '../../src/formats/json.js';
import { GenerateResponseChunk, generateStream } from '../../src/generate.js';
import { Message } from '../../src/message.js';
import { GenerateResponseChunkData, MessageData } from '../../src/model.js';
import { defineProgrammableModel, runAsync } from '../helpers.js';

describe('jsonFormat', () => {
  let registry: Registry;

  beforeEach(() => {
    registry = new Registry();
  });

  const streamingTests = [
    {
      desc: 'parses complete JSON object',
      chunks: [
        {
          text: '{"id": 1, "name": "test"}',
          want: { id: 1, name: 'test' },
        },
      ],
    },
    {
      desc: 'handles partial JSON',
      chunks: [
        {
          text: '{"id": 1',
          want: { id: 1 },
        },
        {
          text: ', "name": "test"}',
          want: { id: 1, name: 'test' },
        },
      ],
    },
    {
      desc: 'handles preamble with code fence',
      chunks: [
        {
          text: 'Here is the JSON:\n\n```json\n',
          want: null,
        },
        {
          text: '{"id": 1}\n```',
          want: { id: 1 },
        },
      ],
    },
  ];

  for (const st of streamingTests) {
    it(st.desc, () => {
      const parser = jsonFormatter.handler();
      const chunks: GenerateResponseChunkData[] = [];

      for (const chunk of st.chunks) {
        const newChunk: GenerateResponseChunkData = {
          index: 0,
          role: 'model',
          content: [{ text: chunk.text }],
        };

        const result = parser.parseChunk!(
          new GenerateResponseChunk(newChunk, {
            index: 0,
            role: 'model',
            previousChunks: [...chunks],
          })
        );
        chunks.push(newChunk);

        assert.deepStrictEqual(result, chunk.want);
      }
    });
  }

  const messageTests = [
    {
      desc: 'parses complete JSON response',
      message: {
        role: 'model',
        content: [{ text: '{"id": 1, "name": "test"}' }],
      },
      want: { id: 1, name: 'test' },
    },
    {
      desc: 'handles empty response',
      message: {
        role: 'model',
        content: [{ text: '' }],
      },
      want: null,
    },
    {
      desc: 'parses JSON with preamble and code fence',
      message: {
        role: 'model',
        content: [{ text: 'Here is the JSON:\n\n```json\n{"id": 1}\n```' }],
      },
      want: { id: 1 },
    },
  ];

  for (const rt of messageTests) {
    it(rt.desc, () => {
      const parser = jsonFormatter.handler();
      assert.deepStrictEqual(
        parser.parseMessage(new Message(rt.message as MessageData)),
        rt.want
      );
    });
  }
});

describe('jsonFormat e2e', () => {
  let registry: Registry;

  beforeEach(() => {
    registry = new Registry();
    configureFormats(registry);
  });

  it('injects the instructions into the request', async () => {
    let pm = defineProgrammableModel(registry);
    pm.handleResponse = async (req, sc) => {
      await runAsync(() => sc?.({ content: [{ text: '```\n{' }] }));
      await runAsync(() => sc?.({ content: [{ text: '"foo": "b' }] }));
      await runAsync(() => sc?.({ content: [{ text: 'ar"' }] }));
      await runAsync(() => sc?.({ content: [{ text: '}\n```"' }] }));
      return await runAsync(() => ({
        message: {
          role: 'model',
          content: [{ text: '```\n{"foo": "bar"}\n```' }],
        },
      }));
    };

    const { response, stream } = await generateStream(registry, {
      model: 'programmableModel',
      prompt: 'generate json',
      output: {
        format: 'json',
        schema: z.object({
          foo: z.string(),
        }),
      },
    });
    const chunks: any = [];
    for await (const chunk of stream) {
      chunks.push(chunk.output);
    }
    assert.deepEqual((await response).output, { foo: 'bar' });
    assert.deepStrictEqual(chunks, [
      {},
      { foo: 'b' },
      { foo: 'bar' },
      { foo: 'bar' },
    ]);
  });
});

File: go/internal/base/json_test.go
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package base

import (
	"testing"

	"github.com/google/go-cmp/cmp"
)

func TestExtractJSONFromMarkdown(t *testing.T) {
	tests := []struct {
		desc string
		in   string
		want string
	}{
		{
			desc: "no markdown",
			in:   "abcdefg",
			want: "abcdefg",
		},
		{
			desc: "no markdown (with line breaks)",
			in:   "ab\ncd\nfg",
			want: "ab\ncd\nfg",
		},
		{
			desc: "simple markdown",
			in:   "```foo bar```",
			want: "```foo bar```",
		},
		{
			desc: "empty markdown",
			in:   "``` ```",
			want: "``` ```",
		},
		{
			desc: "json markdown",
			in:   "```json{\"a\":1}```",
			want: "{\"a\":1}",
		},
		{
			desc: "json multiple line markdown",
			in:   "```json\n{\"a\": 1}\n```",
			want: "{\"a\": 1}",
		},
		{
			desc: "returns first of multiple blocks",
			in:   "```json{\"a\":\n1}```\n```json\n{\"b\":\n1}```",
			want: "{\"a\":\n1}",
		},
		{
			desc: "yaml markdown",
			in:   "```yaml\nkey: 1\nanother-key: 2```",
			want: "```yaml\nkey: 1\nanother-key: 2```",
		},
		{
			desc: "yaml + json markdown",
			in:   "```yaml\nkey: 1\nanother-key: 2``` ```json\n{\"a\": 1}\n```",
			want: "{\"a\": 1}",
		},
		{
			desc: "json + yaml markdown",
			in:   "```json\n{\"a\": 1}\n``` ```yaml\nkey: 1\nanother-key: 2```",
			want: "{\"a\": 1}",
		},
		{
			desc: "uppercase JSON identifier",
			in:   "```JSON\n{\"a\": 1}\n```",
			want: "{\"a\": 1}",
		},
		{
			desc: "mixed case Json identifier",
			in:   "```Json\n{\"a\": 1}\n```",
			want: "{\"a\": 1}",
		},
		{
			desc: "plain code block without identifier",
			in:   "```\n{\"a\": 1}\n```",
			want: "{\"a\": 1}",
		},
		{
			desc: "plain code block with text before",
			in:   "Here is the result:\n\n```\n{\"title\": \"Pizza\"}\n```",
			want: "{\"title\": \"Pizza\"}",
		},
		{
			desc: "json block preferred over plain block",
			in:   "```\n{\"plain\": true}\n``` then ```json\n{\"json\": true}\n```",
			want: "{\"json\": true}",
		},
		{
			desc: "json block with spaces",
			in:   "``` json\n{\"a\": 1}\n```",
			want: "{\"a\": 1}",
		},
		{
			desc: "implicit json block",
			in:   "```{\"a\": 1}```",
			want: "{\"a\": 1}",
		},
		{
			desc: "implicit json block array",
			in:   "```[1, 2]```",
			want: "[1, 2]",
		},
	}
	for _, tc := range tests {
		t.Run(tc.desc, func(t *testing.T) {
			if diff := cmp.Diff(ExtractJSONFromMarkdown(tc.in), tc.want); diff != "" {
				t.Errorf("ExtractJSONFromMarkdown diff (+got -want):\n%s", diff)
			}
		})
	}
}

func TestSchemaAsMap(t *testing.T) {
	type Bar struct {
		Bar string
	}
	type Foo struct {
		BarField Bar
		Str      string
	}

	want := map[string]any{
		"additionalProperties": bool(false),
		"properties": map[string]any{
			"BarField": map[string]any{
				"additionalProperties": bool(false),
				"properties": map[string]any{
					"Bar": map[string]any{"type": string("string")},
				},
				"required": []any{string("Bar")},
				"type":     string("object"),
			},
			"Str": map[string]any{"type": string("string")},
		},
		"required": []any{string("BarField"), string("Str")},
		"type":     string("object"),
	}

	got := SchemaAsMap(InferJSONSchema(Foo{}))
	if diff := cmp.Diff(got, want); diff != "" {
		t.Errorf("SchemaAsMap diff (+got -want):\n%s", diff)
	}
}

func TestSchemaAsMapRecursive(t *testing.T) {
	type Node struct {
		Value    string  `json:"value,omitempty"`
		Children []*Node `json:"children,omitempty"`
	}

	schema := SchemaAsMap(InferJSONSchema(Node{}))

	// With DoNotReference and recursion limiting, the schema should be flat
	// and recursive references should become "any" schema.
	if _, ok := schema["$defs"]; ok {
		t.Error("expected no $defs with DoNotReference: true")
	}

	if _, ok := schema["$ref"]; ok {
		t.Error("expected no $ref with DoNotReference: true")
	}

	// Check top-level structure
	if schema["type"] != "object" {
		t.Errorf("expected type to be object, got %v", schema["type"])
	}

	props, ok := schema["properties"].(map[string]any)
	if !ok {
		t.Fatal("expected properties in schema")
	}

	// Check value field
	valueField, ok := props["value"].(map[string]any)
	if !ok {
		t.Fatal("expected value field in properties")
	}
	if valueField["type"] != "string" {
		t.Errorf("expected value.type to be string, got %v", valueField["type"])
	}

	// Check children field - recursive reference should be "any" schema
	childrenField, ok := props["children"].(map[string]any)
	if !ok {
		t.Fatal("expected children field in properties")
	}
	if childrenField["type"] != "array" {
		t.Errorf("expected children.type to be array, got %v", childrenField["type"])
	}

	items, ok := childrenField["items"].(map[string]any)
	if !ok {
		t.Fatal("expected children to have items")
	}
	// The recursive Node reference should have become an "any" schema
	if items["additionalProperties"] != true {
		t.Errorf("expected children.items to be 'any' schema (additionalProperties: true), got %v", items)
	}
}
